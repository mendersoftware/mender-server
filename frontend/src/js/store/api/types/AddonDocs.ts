/* generated using openapi-typescript-codegen -- do not edit */
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
/**
 * Mender addon definition.
 */
export type AddonDocs = {
  /**
   * Name of the addon (one of ['configure', 'troubleshoot']).
   */
  name?: string;
  /**
   * Enabled/disabled flag.
   */
  enabled?: boolean;
};
