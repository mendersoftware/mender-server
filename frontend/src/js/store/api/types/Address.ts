/* generated using openapi-typescript-codegen -- do not edit */
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
import type { AddressUpdate } from "./AddressUpdate";
export type Address = AddressUpdate;
