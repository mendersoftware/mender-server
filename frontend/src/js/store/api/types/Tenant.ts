/* generated using openapi-typescript-codegen -- do not edit */
/* istanbul ignore file */
/* tslint:disable */
/* eslint-disable */
/**
 * The tenant descriptor.
 */
export type Tenant = {
  /**
   * The id of a tenant.
   */
  id: string;
  /**
   * The name of a tenant.
   */
  name?: string;
  /**
   * The id of a parent tenant.
   */
  parent?: string;
};
