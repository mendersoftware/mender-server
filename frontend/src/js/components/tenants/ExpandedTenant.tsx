// Copyright 2024 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

import { Button, Checkbox, Divider, Drawer, FormControl, FormControlLabel, FormHelperText, TextField, formControlLabelClasses } from '@mui/material';
import { makeStyles } from 'tss-react/mui';

import { TwoColumns } from '@northern.tech/common-ui/ConfigurationObject';
import { ConfirmModal } from '@northern.tech/common-ui/ConfirmModal';
import { DrawerTitle } from '@northern.tech/common-ui/DrawerTitle';
import actions from '@northern.tech/store/actions';
import { generateTenantPathById } from '@northern.tech/store/locationutils';
import { getOrganization, getSsoConfig } from '@northern.tech/store/selectors';
import { AppDispatch } from '@northern.tech/store/store';
import { editTenantDeviceLimit, removeTenant } from '@northern.tech/store/thunks';
import copy from 'copy-to-clipboard';

import { DeviceCount } from '../header/devicecount';
import { Tenant } from './types';

interface ExpandedTenantProps {
  onCloseClick: () => void;
  tenant: Tenant;
}

const useStyles = makeStyles()(theme => ({
  devLimitInput: { minWidth: 150 },
  formWrapper: { display: 'flex', flexDirection: 'column', gap: theme.spacing(2), maxWidth: 750, [`.${formControlLabelClasses.root}`]: { marginTop: 0 } },
  ssoLink: {
    marginLeft: `calc(1em + ${theme.spacing(1.5)})`, // 1em as the width of the checkbox + the padding around the checkbox
    marginTop: theme.spacing(-1)
  }
}));
const { setSnackbar } = actions;

export const ExpandedTenant = (props: ExpandedTenantProps) => {
  const { onCloseClick, tenant } = props;
  const { name, id, device_limit = 0, device_count = 0, binary_delta } = tenant;

  const [shouldDelete, setShouldDelete] = useState<boolean>(false);
  const [newLimitForm, setNewLimitForm] = useState<boolean>(false);
  const [newLimit, setNewLimit] = useState<number>(device_limit);
  const [hasLimitError, setHasLimitError] = useState<boolean>(false);

  const { device_count: spDeviceUtilization, device_limit: spDeviceLimit } = useSelector(getOrganization);
  const ssoConfig = useSelector(getSsoConfig);

  const currentLimit = spDeviceLimit - spDeviceUtilization + device_limit;
  const { classes } = useStyles();
  const dispatch = useDispatch<AppDispatch>();

  const copyLinkToClipboard = () => {
    const location = window.origin + '/ui';
    copy(`${location}${generateTenantPathById(id)}`);
    dispatch(setSnackbar('Link copied to clipboard'));
  };

  const onChangeLimit = ({ target: { validity, value } }) => {
    if (validity.valid) {
      setNewLimit(value);
      return setHasLimitError(false);
    }
    setHasLimitError(true);
  };

  const onNewLimitSubmit = async () => {
    await dispatch(editTenantDeviceLimit({ id, name, newLimit: Number(newLimit) }));
    setNewLimitForm(false);
  };

  const deleteTenant = () => dispatch(removeTenant({ id }));

  return (
    <Drawer onClose={onCloseClick} open={true} PaperProps={{ style: { minWidth: '67vw' } }} anchor="right">
      <DrawerTitle
        title={`Tenant Information for ${name}`}
        onLinkCopy={copyLinkToClipboard}
        preCloser={<Button onClick={() => setShouldDelete(true)}>Delete tenant</Button>}
        onClose={onCloseClick}
      />
      <Divider className="margin-bottom-large" />
      <div className={classes.formWrapper}>
        <TwoColumns className="align-self-start" copyable items={{ name, ID: id }} />
        <FormControlLabel control={<Checkbox color="primary" size="small" disabled checked={binary_delta} />} label="Enable Delta Artifact generation" />
        {!!ssoConfig && (
          <>
            <FormControlLabel
              control={<Checkbox color="primary" size="small" checked disabled />}
              label="Restrict to Service Providerâ€™s Single Sign-On settings"
            />
            <Link className={classes.ssoLink} to="/settings/organization-and-billing">
              View Single Sign-On settings
            </Link>
          </>
        )}
        <div className={`flexbox ${newLimitForm ? '' : 'center-aligned'} margin-top-small`}>
          <DeviceCount current={device_count} max={device_limit} variant="detailed" />
          <div className="margin-left">
            {newLimitForm ? (
              <FormControl className={classes.formWrapper}>
                <div className="flexbox center-aligned">
                  <TextField
                    className={classes.devLimitInput}
                    label="Set device limit"
                    type="number"
                    onChange={onChangeLimit}
                    slotProps={{ htmlInput: { min: device_count, max: currentLimit } }}
                    error={hasLimitError}
                    value={newLimit}
                  />
                  <Button
                    className="margin-left"
                    onClick={() => {
                      setNewLimit(device_limit);
                      setNewLimitForm(false);
                    }}
                  >
                    Cancel
                  </Button>
                  <Button className="margin-left-x-small" onClick={onNewLimitSubmit} color="primary" variant="contained">
                    Save
                  </Button>
                </div>
                <FormHelperText className={`${hasLimitError ? 'warning' : 'info'} margin-top-none`}>Maximum limit: {currentLimit}</FormHelperText>
                <FormHelperText className="info margin-top-none">
                  {spDeviceUtilization} devices assigned of maximum {spDeviceLimit} across all tenants.
                  <br />
                  <a href="mailto:support@mender.io" target="_blank" rel="noopener noreferrer">
                    Contact support
                  </a>{' '}
                  to increase your total limit
                </FormHelperText>
              </FormControl>
            ) : (
              <Button onClick={() => setNewLimitForm(true)}>Edit device limit</Button>
            )}
          </div>
        </div>
        <ConfirmModal
          header="Are you sure you want to delete this tenant?"
          description="All devices, users, artifacts and audit logs associated with the tenant will be removed."
          toType="delete"
          open={shouldDelete}
          close={() => setShouldDelete(false)}
          onConfirm={() => {
            deleteTenant();
            setShouldDelete(false);
            onCloseClick();
          }}
        />
      </div>
    </Drawer>
  );
};
