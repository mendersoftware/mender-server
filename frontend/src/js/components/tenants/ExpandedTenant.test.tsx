// Copyright 2024 Northern.tech AS
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.
import { Tenant } from '@northern.tech/store/api/types/Tenant';
import { getSessionInfo } from '@northern.tech/store/auth';
import { initialState as initialOrganizationState } from '@northern.tech/store/organizationSlice';
import * as OrganizationActions from '@northern.tech/store/thunks';
import { act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import { defaultState, undefineds } from '../../../../tests/mockData';
import { render } from '../../../../tests/setupTests';
import { ExpandedTenant } from './ExpandedTenant';

let state;
let tenant: Tenant;
describe('ExpandedTenant', () => {
  beforeAll(() => {
    state = {
      ...defaultState,
      organization: {
        ...defaultState.organization,
        tenantList: {
          ...initialOrganizationState.tenantList,
          tenants: [
            {
              id: '671a0f1dd58c813118fe8622',
              parent_tenant_id: '6718de64b42e08dea2a2065d',
              name: 'child2',
              tenant_token: 'mQDYRCr-tGbDuJhPp7fArbfTA5htVTWE9G204AzhDUM',
              status: 'active',
              additional_info: {
                marketing: false,
                campaign: ''
              },
              plan: 'enterprise',
              trial: false,
              trial_expiration: null,
              service_provider: false,
              created_at: '2024-10-24T09:10:53.281Z',
              cancelled_at: null,
              children_tenants: null,
              max_child_tenants: 0,
              device_count: 0,
              device_limit: 100,
              binary_delta: true
            }
          ]
        },
        organization: {
          ...defaultState.organization.organization,
          device_count: 20,
          device_limit: 200
        },
        users: { ...defaultState.users, currentSession: getSessionInfo() }
      }
    };
    tenant = state.organization.tenantList.tenants[0];
  });
  it('renders correctly', () => {
    const { baseElement } = render(<ExpandedTenant onCloseClick={vi.fn} tenant={tenant} />, {
      preloadedState: state
    });
    const view = baseElement;
    expect(view).toMatchSnapshot();
    expect(view).toEqual(expect.not.stringMatching(undefineds));
  });
  it('works as intended', async () => {
    const user = userEvent.setup({ advanceTimers: vi.advanceTimersByTime });
    const editDeviceLimit = vi.spyOn(OrganizationActions, 'editTenantDeviceLimit');

    const newLimit = '5';
    render(<ExpandedTenant onCloseClick={vi.fn} tenant={tenant} />, { preloadedState: state });
    expect(screen.queryByText(`Tenant Information for ${tenant.name}`));
    await user.click(screen.getByRole('button', { name: /edit device limit/i }));
    const limitInput = screen.getByLabelText(/set device limit/i);
    await user.clear(limitInput);
    await user.type(limitInput, newLimit);
    await user.click(screen.getByRole('button', { name: /save/i }));
    await act(async () => {
      vi.runOnlyPendingTimers();
      vi.runAllTicks();
    });
    expect(editDeviceLimit).toHaveBeenCalledWith({ newLimit: Number(newLimit), name: tenant.name, id: tenant.id });
  });
});
