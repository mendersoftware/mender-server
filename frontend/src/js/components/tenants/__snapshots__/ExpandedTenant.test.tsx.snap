// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandedTenant renders correctly 1`] = `
.emotion-0 {
  position: fixed;
  z-index: 1300;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  z-index: 1200;
}

.emotion-1 {
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  right: 0;
  bottom: 0;
  top: 0;
  left: 0;
  background-color: rgba(0, 0, 0, 0.5);
  -webkit-tap-highlight-color: transparent;
  z-index: -1;
}

.emotion-2 {
  background-color: #fff;
  color: rgba(10, 10, 11, 0.78);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow: var(--Paper-shadow);
  background-image: var(--Paper-overlay);
  overflow-y: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  z-index: 1200;
  -webkit-overflow-scrolling: touch;
  position: fixed;
  top: 0;
  outline: 0;
  right: 0;
  min-width: 40vw;
  max-width: 80vw;
  padding: 30px 75px 5%;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.3929rem;
  padding: 8px;
  border-radius: 50%;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  --IconButton-hoverBg: rgba(0, 0, 0, 0.04);
  padding: 12px;
  font-size: 1.625rem;
  font-size: 1.2rem;
  color: rgba(0, 0, 0, 0.54);
}

.emotion-3::-moz-focus-inner {
  border-style: none;
}

.emotion-3.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-3 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-3:hover {
  background-color: var(--IconButton-hoverBg);
}

@media (hover: none) {
  .emotion-3:hover {
    background-color: transparent;
  }
}

.emotion-3.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3.MuiIconButton-loading {
  color: transparent;
}

.emotion-4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.3929rem;
}

.emotion-4 iconButton {
  margin-right: 8px;
}

.emotion-5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: Lato,sans-serif;
  font-weight: 500;
  font-size: 0.8125rem;
  line-height: 1.75;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 16px;
  border: 0;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  padding: 6px 8px;
  color: var(--variant-textColor);
  background-color: var(--variant-textBg);
  --variant-textColor: #337a87;
  --variant-outlinedColor: #337a87;
  --variant-outlinedBorder: rgba(51, 122, 135, 0.5);
  --variant-containedColor: #fff;
  --variant-containedBg: #337a87;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 2px;
  font-size: 14px;
  font-weight: bold;
  padding: 10px 15px;
}

.emotion-5::-moz-focus-inner {
  border-style: none;
}

.emotion-5.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-5 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-5:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-5.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

@media (hover: hover) {
  .emotion-5:hover {
    --variant-containedBg: rgb(35, 85, 94);
    --variant-textBg: rgba(51, 122, 135, 0.04);
    --variant-outlinedBorder: #337a87;
    --variant-outlinedBg: rgba(51, 122, 135, 0.04);
  }
}

.emotion-5.MuiButton-loading {
  color: transparent;
}

.emotion-5:hover {
  colors: #337a87;
}

.emotion-5.MuiButton-text {
  color: rgba(10, 10, 11, 0.78);
}

.emotion-8 {
  margin: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0, 0, 0, 0.12);
  border-bottom-width: thin;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 16px;
  max-width: 750px;
}

.emotion-9 .MuiFormControlLabel-root {
  margin-top: 0;
}

.emotion-10 .key>b {
  background-color: #f7f7f7;
  color: rgba(0, 0, 0, 0.87);
}

.emotion-11 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.1607rem;
  color: #337a87;
}

.emotion-11 iconButton {
  margin-right: 8px;
}

.emotion-13 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  vertical-align: middle;
  -webkit-tap-highlight-color: transparent;
  margin-left: -11px;
  margin-right: 16px;
  margin-top: 18px;
}

.emotion-13.Mui-disabled {
  cursor: default;
}

.emotion-13 .MuiFormControlLabel-label.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-14 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  padding: 9px;
  border-radius: 50%;
  color: rgba(0, 0, 0, 0.54);
  color: rgba(10, 10, 11, 0.78);
}

.emotion-14::-moz-focus-inner {
  border-style: none;
}

.emotion-14.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-14 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-14:hover {
  background-color: rgba(51, 122, 135, 0.04);
}

.emotion-14.Mui-checked,
.emotion-14.MuiCheckbox-indeterminate {
  color: #337a87;
}

.emotion-14.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

@media (hover: none) {
  .emotion-14:hover {
    background-color: transparent;
  }
}

.emotion-15 {
  cursor: inherit;
  position: absolute;
  opacity: 0;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
  z-index: 1;
}

.emotion-16 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  fill: currentColor;
  font-size: 1.1607rem;
}

.emotion-16 iconButton {
  margin-right: 8px;
}

.emotion-17 {
  margin: 0;
  line-height: 1.5;
  font-family: Lato,sans-serif;
  font-weight: 400;
  font-size: 0.9286rem;
}

.emotion-18 {
  width: 375px;
}

.emotion-19 {
  margin: 11px 11px 11px 0;
}

.emotion-20 {
  background: #e6f2f1;
  color: #969696;
  border-radius: 2px;
  padding: 4px 6px;
}

.emotion-21 {
  position: relative;
  overflow: hidden;
  display: block;
  height: 4px;
  z-index: 0;
  background-color: rgb(177, 204, 209);
  background-color: #e9e9e9;
}

@media print {
  .emotion-21 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-22 {
  width: 100%;
  position: absolute;
  left: 0;
  bottom: 0;
  top: 0;
  -webkit-transition: -webkit-transform 0.2s linear;
  transition: transform 0.2s linear;
  transform-origin: left;
  background-color: #337a87;
  -webkit-transition: -webkit-transform .4s linear;
  transition: transform .4s linear;
}

<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDrawer-root MuiDrawer-anchorRight MuiDrawer-modal MuiModal-root emotion-0"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop emotion-1"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      aria-modal="true"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation16 MuiDrawer-paper MuiDrawer-paperAnchorRight emotion-2"
      role="dialog"
      style="--Paper-shadow: 0px 8px 10px -5px rgba(0,0,0,0.2),0px 16px 24px 2px rgba(0,0,0,0.14),0px 6px 30px 5px rgba(0,0,0,0.12); min-width: 67vw;"
      tabindex="-1"
    >
      <div
        class="flexbox margin-bottom-small space-between"
      >
        <div
          class="flexbox center-aligned"
        >
          <h3
            class="capitalized-start flexbox center-aligned"
          >
            Tenant Information for child2
          </h3>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge emotion-3"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-4"
              data-testid="LinkIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5"
              />
            </svg>
          </button>
        </div>
        <div
          class="flexbox center-aligned"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary emotion-5"
            tabindex="0"
            type="button"
          >
            Delete tenant
          </button>
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge emotion-3"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-4"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth margin-bottom-large emotion-8"
      />
      <div
        class="emotion-9"
      >
        <div
          class="break-all two-columns emotion-10  align-self-start"
        >
          <div
            class="align-right key muted"
          >
            <b>
              name
            </b>
          </div>
          <div
            class="flexbox clickable"
          >
            <div
              title="child2"
            >
              child2
            </div>
            <svg
              aria-hidden="true"
              aria-label="Copy to clipboard"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall margin-left-small fadeOut emotion-11"
              data-mui-internal-clone-element="true"
              data-testid="FileCopyOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11zM8 21V7h6v5h5v9z"
              />
            </svg>
          </div>
          <div
            class="align-right key muted"
          >
            <b>
              ID
            </b>
          </div>
          <div
            class="flexbox clickable"
          >
            <div
              title="671a0f1dd58c813118fe8622"
            >
              671a0f1dd58c813118fe8622
            </div>
            <svg
              aria-hidden="true"
              aria-label="Copy to clipboard"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeSmall margin-left-small fadeOut emotion-11"
              data-mui-internal-clone-element="true"
              data-testid="FileCopyOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12zm-1 4H8c-1.1 0-1.99.9-1.99 2L6 21c0 1.1.89 2 1.99 2H19c1.1 0 2-.9 2-2V11zM8 21V7h6v5h5v9z"
              />
            </svg>
          </div>
        </div>
        <label
          class="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd emotion-13"
        >
          <span
            aria-disabled="true"
            class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall Mui-disabled PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall Mui-checked Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeSmall emotion-14"
            tabindex="-1"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input emotion-15"
              data-indeterminate="false"
              disabled=""
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall emotion-16"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled emotion-17"
          >
            Enable Delta Artifact generation
          </span>
        </label>
        <div
          class="flexbox center-aligned margin-top-small"
        >
          <div
            class="emotion-18"
          >
            <div
              class="flexbox centered space-between margin-left-x-small margin-right-x-small"
            >
              <div
                class="flexbox centered"
              >
                <p
                  class="emotion-19"
                >
                  Devices: 
                  0
                  /
                  100
                </p>
              </div>
              <div
                class="emotion-20"
              >
                100
                 devices left
              </div>
            </div>
            <span
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="0"
              class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate margin-left-x-small margin-right-x-small emotion-21"
              role="progressbar"
            >
              <span
                class="MuiLinearProgress-bar MuiLinearProgress-bar1 MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate emotion-22"
                style="transform: translateX(-100%);"
              />
            </span>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth margin-top-small emotion-8"
            />
          </div>
          <div
            class="margin-left"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary emotion-5"
              tabindex="0"
              type="button"
            >
              Edit device limit
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
